### PR TITLE
[SPARK-20461][Core][SS]Use UninterruptibleThread for Executor and fix the potential hang in CachedKafkaConsumer

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

This PR changes Executor's threads to `UninterruptibleThread` so that we can use `runUninterruptibly` in `CachedKafkaConsumer`. However, this is just best effort to avoid hanging forever. If the user uses`CachedKafkaConsumer` in another thread (e.g., create a new thread or Future), the potential hang may still happen.

## How was this patch tested?

The new added test.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 1]
```diff
@@ -84,7 +86,20 @@ private[spark] class Executor(
...
+  private val threadPool = {
+    val threadFactory = new ThreadFactoryBuilder()
+      .setDaemon(true)
+      .setNameFormat("Executor task launch worker-%d")
+      .setThreadFactory(new ThreadFactory {
+        override def newThread(r: Runnable): Thread =
+          // Use UninterruptibleThread to run tasks so that we can allow running codes without being
+          // interrupted by `Thread.interrupt()`. Some issues, such as KAFKA-1894, HADOOP-10622,
+          // will hang forever if some methods are interrupted.
+          new UninterruptibleThread(r, "unused") // thread name will be set by ThreadFactoryBuilder
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: Mostly of the codes are copied from `ThreadUtils`. This one is the only difference that matters.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaConsumer.scala [hunk 2]
```diff
@@ -62,11 +63,20 @@ private[kafka010] case class CachedKafkaConsumer private(
 
   case class AvailableOffsetRange(earliest: Long, latest: Long)
 
+  private def runUninterruptiblyIfPossiable[T](body: => T): T = Thread.currentThread match {
```
Comment (tdas) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaConsumer.scala]: Possiable -> Possible

Comment (kunalkhamar) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaConsumer.scala]: nit: rename `runUninterruptiblyIfPossible`

**Comment by mridulm** [comment]
Task interruption is part of our API iirc - kill with interrupt = true.
Doing this will break that contract - did I miss something here ?

Also, IIRC @joshrosen had done a bunch of work regarding task reaper, etc - which should get affected by this.

**Comment by zsxwing** [comment]
@mridulm this only affect codes calling `runUninterruptibly` which is not a public API, so won't break any existing codes. The worst case of this PR is some task needs to wait until network timeout. However, if removing `runUninterruptibly` added in this PR, it's pretty easy to be stuck in an endless loop of Kafka client when interrupting it. That's wasting resources and unacceptable.

**Comment by mridulm** [comment]
@zsxwing Got it, thanks for clarifying.

**Comment by tdas** [comment]
LGTM. Merging this to master and 2.2

**Pull request closed without merging**